Packaging: Add vagrant tests for windows service

rjernst opened the pull request:
This commit adds tests using the vagrant windows image for the windows
service manager script.

rjernst committed:
Tests: Add vagrant tests for windows service

This commit adds tests using the vagrant windows image for the windows
service manager script.

rjernst committed:
Merge branch 'master' into launcher2

rjernst committed:
remove debug printout

rjernst committed:
add windows assumption

rjernst commented:
@elasticmachine test this please

andyb-elastic commented on qa/vagrant/src/main/java/org/elasticsearch/packaging/test/ArchiveTestCase.java:
```diff
@@ -69,25 +72,42 @@
+        final Result ossConflicts = sh.run("rpm -qp --conflicts " + getDistributionFile(OSS_RPM));
+
+        TestCase.assertTrue(Pattern.compile("(?m)^elasticsearch-oss\\s*$").matcher(defaultConflicts.stdout).find());
+        TestCase.assertTrue(Pattern.compile("(?m)^elasticsearch\\s*$").matcher(ossConflicts.stdout).find());
```
If we're relocating these tests into a `PackagingTestCase` subclass would it make more sense for them to go in `PackageTestCase` (I now see the issue with these names) rather than `ArchiveTestCase`

rjernst replied:
I don't think they make sense at the PackagingTestCase level, since they are specific to rpm/deb. I moved them into the oss rpm/deb tests, since they only need to run once.

andyb-elastic commented on qa/vagrant/src/main/java/org/elasticsearch/packaging/test/PackagingTestCase.java:
```diff
@@ -21,31 +21,54 @@
  * familiar with from {@link org.elasticsearch.test.ESTestCase} without having to apply its behavior that's not relevant here
  */
-public abstract class PackagingTestCase {
+public abstract class PackagingTestCase extends Assert {
```
What's the motivation for extending `Assert`?

rjernst replied:
It makes assert methods available without needing to do imports. This is what LuceneTestCase does (and by extension, ESTestCase).

andyb-elastic commented on qa/vagrant/src/main/java/org/elasticsearch/packaging/test/PackagingTestCase.java:
```diff
@@ -21,31 +21,54 @@
+    }
+
+    /** The {@link Distribution} that should be tested in this case */
+    protected abstract Distribution distribution();
```
If we're adding this here, can we remove the abstract method declarations from `ArchiveTestCase` and `PackageTestCase`

rjernst replied:
I don't see it in ArchiveTestCase anymore, but I've removed it now from PackageTestCase.

andyb-elastic commented on distribution/src/bin/elasticsearch-service.bat:
```diff
@@ -34,6 +34,7 @@ if /i %SERVICE_CMD% == start goto doStart
 if /i %SERVICE_CMD% == stop goto doStop
 if /i %SERVICE_CMD% == manager goto doManagment
 echo Unknown option "%SERVICE_CMD%"
+exit /B 1
```
What's the motivation behind `/B` here?

>    /B        When used in a batch script, this option will exit 
              only the script (or subroutine) but not CMD.EXE
              If executed from outside a batch script, if will quit CMD.exe

rjernst replied:
I copied what we had for existing exit calls. Before we would exist 0 even though there was an error in the cases I have added explicit exit.

andyb-elastic commented on qa/vagrant/src/main/java/org/elasticsearch/packaging/test/WindowsServiceTestCase.java:
```diff
@@ -0,0 +1,209 @@
+    @After
+    public void uninstallService() {
+        sh.runIgnoreExitCode(serviceScript + " remove");
+    }
```
I see why this needs to be in an `@After` method here. I think it would also make sense to do this in the routine that cleans everything up between tests https://github.com/elastic/elasticsearch/blob/2aba52de8f9315b0e384e1c657d7b0401d26a1b0/qa/vagrant/src/main/java/org/elasticsearch/packaging/util/Cleanup.java#L90

rjernst replied:
The call to the cleanup method (or in the cleanup method) is outside the windows service test case, which is the only thing that knows about the service existing.

andyb-elastic commented on qa/vagrant/src/main/java/org/elasticsearch/packaging/test/WindowsServiceTestCase.java:
```diff
@@ -0,0 +1,209 @@
+        assertThat(result.stdout, containsString("Unknown option \"bogus\""));
+    }
+
+    private void assertService(String id, String status, String displayName) {
```
Nit: can you put this and `runWithoutJava` next to each other

rjernst replied:
done

andyb-elastic commented on qa/vagrant/src/main/java/org/elasticsearch/packaging/test/WindowsServiceTestCase.java:
```diff
@@ -0,0 +1,209 @@
+    public void test12InstallService() {
+        sh.run(serviceScript + " install");
+        assertService(DEFAULT_ID, "Stopped", DEFAULT_DISPLAY_NAME);
+        sh.run(serviceScript + " remove");
```
I think these should get taken care of by `uninstallService`? Or is the point here that we want to assert exit code 0 when uninstalling it in these tests

rjernst replied:
Yeah the point is we want exit code 0 here.

andyb-elastic commented on qa/vagrant/src/main/java/org/elasticsearch/packaging/test/WindowsServiceTestCase.java:
```diff
@@ -0,0 +1,209 @@
+    public void test30StartStop() throws IOException {
+        sh.run(serviceScript + " install");
+        sh.run(serviceScript + " start");
+        ServerUtils.waitForElasticsearch();
```
Usually we also make a few API calls to the server, e.g. https://github.com/elastic/elasticsearch/blob/2aba52de8f9315b0e384e1c657d7b0401d26a1b0/qa/vagrant/src/main/java/org/elasticsearch/packaging/test/PackageTestCase.java#L121-L122

I'm not completely sold on the value of those though

rjernst replied:
Sure, I've added calling the same test method here.

andyb-elastic reviewed (changes requested):
Looks good, just a couple minor suggestions and questions

rjernst committed:
Merge branch 'master' into launcher2

rjernst committed:
address PR feedback

rjernst commented:
@andyb-elastic I believe I have addressed all your comments.

andyb-elastic commented:
lgtm

rjernst committed:
Merge branch 'master' into launcher2

rjernst committed:
Merge branch 'master' into launcher2

rjernst committed:
remove unused import

rjernst committed:
Merge branch 'master' into launcher2

rjernst committed:
Merge branch 'master' into launcher2

rjernst committed:
Merge branch 'master' into launcher2

rjernst committed:
Merge branch 'master' into launcher2

rjernst committed:
Merge branch 'master' into launcher2

rjernst committed:
test debugging improvements

rjernst committed:
Merge branch 'master' into launcher2

rjernst committed:
Merge branch 'master' into launcher2

rjernst committed:
Merge branch 'master' into launcher2

rjernst committed:
make service start/stop test use service checks

rjernst committed:
fix tests, works with updated apache procrun exes

rjernst committed:
Packaging: Update procrun executables to version 1.1.0

This commit updates the procrun manager and service exes to 1.1.0. There
are a few bug fixes, including for a bug which can cause lingering
processes when removing the service.

rjernst committed:
Merge branch 'master' into launcher2

rjernst committed:
Merge branch 'procrun1.1' into launcher2

rjernst committed:
Merge branch 'master' into launcher2

rjernst committed:
add debugging stuff

rjernst committed:
use logger

rjernst committed:
loop waiting on mystery process

rjernst committed:
try blanket sleep...

rjernst committed:
Merge branch 'master' into launcher2

rjernst committed:
Merge branch 'master' into launcher2

rjernst committed:
temporarily add procdump to get info about lingering process

rjernst commented:
@elasticmachine run packaging tests

rjernst committed:
Merge branch 'master' into launcher2

rjernst committed:
try different flags

rjernst committed:
Merge branch 'master' into launcher2

rjernst committed:
no process copy...

rjernst committed:
try using sc exe

rjernst committed:
"better" logging...

rjernst committed:
turn procdump back on

rjernst committed:
Merge branch 'master' into launcher2

rjernst committed:
more debugging

rjernst committed:
Revert "Packaging: Update procrun executables to version 1.1.0 (#35147)"

This reverts commit bfaa94f53e5b2cf03593427d33b9ee583229373b.

rjernst committed:
isolate windows tests and sleep waiting for zombie

rjernst committed:
reenable other tests

rjernst committed:
cleanup

rjernst committed:
Revert "Revert "Packaging: Update procrun executables to version 1.1.0 (#35147)""

This reverts commit 97e3476c09c821aa88a1001c5e6c31b09d6a8516.

rjernst committed:
re-enable other packaging tests

rjernst committed:
remove temp binary

rjernst committed:
remove commented out code
